File: app/src/main/java/com/example/sw0b_001/Models/EncryptedContent/MessagesRecyclerAdapter.java
package com.example.sw0b_001.Models.EncryptedContent;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.AsyncListDiffer;
import androidx.recyclerview.widget.RecyclerView;

import com.example.sw0b_001.Models.Platforms.Platforms;
import com.example.sw0b_001.Models.Platforms._PlatformsHandler;
import com.example.sw0b_001.Modules.Helpers;
import com.example.sw0b_001.R;
import com.google.android.material.card.MaterialCardView;

import org.jetbrains.annotations.NotNull;

public class MessagesRecyclerAdapter extends RecyclerView.Adapter<MessagesRecyclerAdapter.ViewHolder> {
    public final AsyncListDiffer<EncryptedContent> mDiffer = new AsyncListDiffer(this,
            EncryptedContent.DIFF_CALLBACK);

    @NonNull
    @NotNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull @NotNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.layout_cardlist_recents, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull @NotNull ViewHolder holder, int position) {
        EncryptedContent encryptedContent = mDiffer.getCurrentList().get(position);
        holder.bind(encryptedContent);
    }

    @Override
    public int getItemCount() {
        return this.mDiffer.getCurrentList().size();
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {
        TextView encryptedTextSnippet;
        TextView date, subject;
        ImageView platformLogo;
        MaterialCardView card;

        public ViewHolder(@NonNull @NotNull View itemView) {
            super(itemView);
            card = itemView.findViewById(R.id.recents_card_layout);
            date = itemView.findViewById(R.id.recent_date);
            encryptedTextSnippet = itemView.findViewById(R.id.encryptedTextSnippet);
            subject = itemView.findViewById(R.id.homepage_subject);
            platformLogo = itemView.findViewById(R.id.recents_platform_logo);
        }

        public void bind(EncryptedContent encryptedContent) {
            encryptedTextSnippet.setText(encryptedContent.getEncryptedContent());

            Platforms platforms = _PlatformsHandler.getPlatform(itemView.getContext(),
                    encryptedContent.getPlatformName());

            try {
                platformLogo.setImageResource(_PlatformsHandler.hardGetLogoByName(platforms.getName()));

                String dateStr = Helpers.INSTANCE.formatDate(itemView.getContext(),
                        encryptedContent.getDate());

                date.setText(dateStr);
                subject.setText("Sample Subject");
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }

}

File: app/src/main/java/com/example/sw0b_001/Models/Messages/MessagesRecyclerAdapter.kt
package com.example.sw0b_001.Models.Messages

import android.graphics.BitmapFactory
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import androidx.lifecycle.MutableLiveData
import androidx.recyclerview.widget.AsyncListDiffer
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.example.sw0b_001.Modals.PlatformComposers.ComposeHandlers
import com.example.sw0b_001.Models.Platforms.AvailablePlatforms
import com.example.sw0b_001.Models.Platforms.Platforms
import com.example.sw0b_001.Models.Platforms.StoredPlatformsEntity
import com.example.sw0b_001.Models.Platforms._PlatformsHandler
import com.example.sw0b_001.Modules.Helpers.formatDate
import com.example.sw0b_001.R
import com.google.android.material.card.MaterialCardView
import com.google.android.material.textview.MaterialTextView

class MessagesRecyclerAdapter(private val availablePlatforms: List<AvailablePlatforms>) :
    RecyclerView.Adapter<MessagesRecyclerAdapter.ViewHolder>() {
    val mDiffer: AsyncListDiffer<EncryptedContent> = AsyncListDiffer(this,
            EncryptedContent.DIFF_CALLBACK)

    var messageOnClickListener: MutableLiveData<EncryptedContent> =
        MutableLiveData()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(parent.context)
                .inflate(R.layout.layout_cardlist_recents, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val encryptedContent = mDiffer.currentList[position]
        availablePlatforms.forEach {
            if(it.name == encryptedContent.platformName) {
                holder.bind(encryptedContent, it)
                return@forEach
            }
        }
        holder.card.setOnClickListener {
            messageOnClickListener.value = encryptedContent
        }
    }

    override fun getItemCount(): Int {
        return mDiffer.currentList.size
    }

    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val card: MaterialCardView = itemView.findViewById(R.id.recents_card_layout)
        private val date = itemView.findViewById<MaterialTextView>(R.id.encryptedTextDate)
        private val body = itemView.findViewById<MaterialTextView>(R.id.encryptedTextSnippet)
        private val subject = itemView.findViewById<MaterialTextView>(R.id.homepage_subject)
        private val recipient = itemView.findViewById<MaterialTextView>(R.id.homepage_recipient)
        private val platformLogo = itemView.findViewById<ImageView>(R.id.recents_platform_logo)

        fun bind(messages: EncryptedContent, platforms: AvailablePlatforms) {
            val decomposed = ComposeHandlers.decompose(messages.getEncryptedContent(), platforms)

            body.text = decomposed.body

            val dateStr = formatDate(itemView.context, messages.getDate())
            date.text = dateStr

            try {
                if (platforms.logo != null) {
                    val bitmap = BitmapFactory.decodeByteArray( platforms.logo, 0,
                        platforms.logo!!.size )
                    platformLogo.setImageBitmap(bitmap)
                }

                when(platforms.service_type!!) {
                    Platforms.Type.EMAIL.type -> {
                        subject.text = decomposed.subject
                        recipient.text = decomposed.recipient

                        subject.visibility = View.VISIBLE
                        recipient.visibility = View.VISIBLE
                    }
                    Platforms.Type.MESSAGE.type -> {
                        subject.visibility = View.VISIBLE
                        subject.text = decomposed.subject

                        recipient.visibility = View.GONE
                    } else -> {
                        subject.visibility = View.GONE
                        recipient.visibility = View.GONE
                    }
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }

        }
    }
}
